### PR TITLE
Concurrency utilities

### PR DESCRIPTION
**Pull request by pizzi80**

### CODE REVIEW

**Comment by BalusC** [comment]
Please check javadoc of Utils class: https://github.com/omnifaces/omnifaces/blob/51e243b0c593f9eb3f64a2fe303a8fae35dca940/src/main/java/org/omnifaces/util/Utils.java#L102

**Comment by pizzi80** [comment]
Yes but it's needed for future PR(s) of things that I would like to backport from my forked version

You asked for smaller PR(s), so I've splitted things, but It's time consuming and this was the easiest to backport ...

**Comment by BalusC** [comment]
The idea is if you need to reuse exactly the same private static method from a different class, then it's OK to move it into Utils class. Right now it's used in only one class, the very one where it's being declared.

**Comment by BalusC** [comment]
By the way, take a look at https://github.com/omnifaces/omniutils

A long time ago I've played with the idea to simply split/refactor all of OmniFaces "internal" util classes over there and adjust the OmniFaces build to shadow specific OmniUtils classes so we still end up with a single final JAR, but I couldn't get the build right. Perhaps it's time to take a second look.

cc: @arjantijms

**Comment by pizzi80** [comment]
This is also a very big problem in Jakarta EE "modules"

looking at the code of some modules I saw a lot of copy pasted methods
and Utils class .... 

OmniUtils could teorically be a Jakarta Utils standard module

**Comment by pizzi80** [comment]
> The idea is if you need to reuse exactly the same private static method from a different class, then it's OK to move it into Utils class. Right now it's used in only one class, the very one where it's being declared.

After some months or years, or if another commiter will need those method,
no one will remember that inside a class there is exaclty that utility method

Also no one will take care to make a proper refactor because:
   1) it will increase the number of classes modified with a single PR
   2) fear of bugs
   3) fear of git conflict with other PRs

my 2 cents ;)

**Comment by pizzi80** [comment]
I've reopen this PR to avoid copy/paste on multiple classes

File: src/main/java/org/omnifaces/util/Utils.java [hunk 1]
```diff
@@ -832,6 +834,55 @@ public static void forEach(Object object, Consumer<? super Object> action) {
...
+    // Concurrency ----------------------------------------------------------------------------------------------------
+
+    /**
+     * A {@link FunctionalInterface} to be used with {@link Utils#execAtomic(Lock, Action)}
+     * @since 4.5
+     */
+    @FunctionalInterface
+    public interface Action {
+        void execute() throws Exception;
+    }
```
Comment (BalusC) [inline src/main/java/org/omnifaces/util/Utils.java:846]: I introduced FunctionalInterfaces in 4.6 https://github.com/omnifaces/omnifaces/commit/02f87b90703546cb062f7a4299acc593c13853e5. Please use ThrowingRunnable instead.

File: src/main/java/org/omnifaces/util/Utils.java [hunk 2]
```diff
@@ -832,6 +834,55 @@ public static void forEach(Object object, Consumer<? super Object> action) {
...
+    public static <R> R execAtomic(Lock lock, Supplier<R> task) {
+        lock.lock();
+
+        try {
+            return task.get();
+        }
+        finally {
+            lock.unlock();
+        }
+    }
```
Comment (BalusC) [inline src/main/java/org/omnifaces/util/Utils.java:884]: Not terribly sure on this ..

**Comment by BalusC** [comment]
Closing off in favor of https://github.com/omnifaces/omnifaces/commit/7ddf5ef472ca7e30d8746644c8436d496c6f2b46

File: src/main/java/org/omnifaces/util/Utils.java (hunk 1 above)
Comment (pizzi80) [reply src/main/java/org/omnifaces/util/Utils.java:846]: we could also use the java lang Runnable which is a FunctionalInterfaces and remove the cath block ...

**Comment by pizzi80** [comment]
ok! 

I'll adapt the other PR!

**Pull request closed without merging**